### DISCUSSION TITLE
Rust inlay hints do not work consistently

### DISCUSSION THREAD

**Discussion by JosiahParry** [zed-industries/zed#10390] [Feature Requests]
The one thing that keeps me coming back to try Zed is how nice and clean the editor feels. However, the one thing that sends me away each and every time is the inlay hint support for Rust.

Rust is so tough to write without inlay hints. It makes a world of a difference, plus being able to get match arm suggestions from the rust analyzer makes it so much easier to write. 

In one project, inlay hints work. Except sometimes they'll just turn off? 
<img width="842" alt="image" src="https://github.com/zed-industries/zed/assets/13534508/fc5a7da0-109d-4cf6-8e6a-490f727b465a">

In another project, inlay hints will never appear no amount of changing the toggle will make them appear.

<img width="760" alt="image" src="https://github.com/zed-industries/zed/assets/13534508/0d2316f2-cad0-4802-a42b-622a03a426c6">

Additionally, there is an imperceptible visual cue that inlays are toggled on or off 
<img width="37" alt="image" src="https://github.com/zed-industries/zed/assets/13534508/b19c3bfd-4497-4dab-8cce-28648319d2c4">
<img width="47" alt="image" src="https://github.com/zed-industries/zed/assets/13534508/a5f31e8e-5c4d-4329-9eb9-2c8fe411e79e">

I _really_ want to keep using Zed for my rust coding but the lack of inlay hints and the goods that come along for the ride with the rust analyzer are a bit of a deal breaker :(

**Comment by SomeoneToIgnore** [comment]
Do you have any project examples and repro steps other people can try and debug?

**Reply by JosiahParry** [reply]
Sure! 
The first screenshot with inlay hints working is from https://github.com/josiahParry/arrow-extendr the second screenshot is from https://github.com/JosiahParry/serde_esri

**Reply by SomeoneToIgnore** [reply]
Thanks, the first project does indeed show the hints for me, but I've failed to do anything to make them disappear abruptly — do you have any sequence of actions in your mind?
In general, Zed is somewhat fast to update inlay hints, and what could happen is when a syntax and/or types are broken for the whole block might get spoiled due to that.

Regular:
<img width="1111" alt="image" src="https://github.com/zed-industries/zed/assets/2690773/c93f1aaf-020b-43c8-bc1a-aab75c332224">

With the syntax error:
<img width="1017" alt="image" src="https://github.com/zed-industries/zed/assets/2690773/d8ddbea6-9221-480f-9965-6239030a4b4e">

this is rust-analyzer's choice and Zed barely can influence that with the way inlay hints are modeled now, invalidating after every edit.

----------

The 2nd project is indeed somewhat special, due to the fact that part of its *.rs files being hidden behind the feature.
VSCode does a good job and shows that instantly, and on every edit: 
<img width="1728" alt="image" src="https://github.com/zed-industries/zed/assets/2690773/ebb156ef-898e-418d-8017-80d626832b96">

Zed seems not to know anything about that, and neither I see any related logs on that.

In the included modules, there's nothing that really makes r-a to emit any hints.
If I add a let binding into the `geometry.rs`, I see the hints:
<img width="779" alt="image" src="https://github.com/zed-industries/zed/assets/2690773/9d3b109e-8a0f-4300-8da9-cb1d2a33053c">

----------

So, seems that what Zed lacks is a better detection of non-included files.
LSP server should provide that, but this place is odd: the message from the screenshot with VSCode comes from VSCode's plugin (client part) for r-a: https://github.com/rust-lang/rust-analyzer/blob/657b33b0cb9bd49085202e91ad5b4676532c9140/editors/code/src/client.ts#L154

As I've mentioned, there were nothing in LSP logs (`debug: open language server logs`) command in Zed's command palette, related to non-incuded files, so it seems that the fix is not very straightforwards and might have to start in https://github.com/rust-lang/rust-analyzer/

**Reply by SomeoneToIgnore** [reply]
Or, sorry, reading the client code more, there has to be a diagnostic coming for that unlinked file?
Again, I do not see such diagnostics in the LSP logs (diagnostics are empty for the file), so interesting why r-a omits what we need.

**Reply by JosiahParry** [reply]
Thank you for looking into this so thoroughly! I'll try paying attention to when inlay hints turn off and see if i can reliably repro

**Reply by asnimansari** [reply]
@SomeoneToIgnore what fonts are these?

**Reply by SomeoneToIgnore** [reply]
`monocraft`

**Comment by pronebird** [comment]
How did you turn on type hints with zed?

**Reply by SomeoneToIgnore** [reply]
That depends on the language (and LSP backing it): the Rust ones are turned on by default and should be shown in any Rust project, same for TypeScript, Go and maybe others.

There is also a related section in the default settings:
```json5
// Inlay hint related settings
  "inlay_hints": {
    // Global switch to toggle hints on and off, switched off by default.
    "enabled": false,
    // Toggle certain types of hints on and off, all switched on by default.
    "show_type_hints": true,
    "show_parameter_hints": true,
    // Corresponds to null/None LSP hint type value.
    "show_other_hints": true,
    // Time to wait after editing the buffer, before requesting the hints,
    // set to 0 to disable debouncing.
    "edit_debounce_ms": 700,
    // Time to wait after scrolling the buffer, before requesting the hints,
    // set to 0 to disable debouncing.
    "scroll_debounce_ms": 50
  },
```

**Reply by pronebird** [reply]
@SomeoneToIgnore thanks. Somehow they weren't turned on by default for Rust and when I touched config, all of a sudden, Zed started downloading rust-analyzer 😆

**Comment by wiiznokes** [comment]
https://github.com/zed-industries/zed/issues/17350

**Reply by JosiahParry** [reply]
I know how to turn on the inlay hints. The problem is that they turn themself off very frequently so I'm smashing the combo to turn them back on.

**Comment by celaltas** [comment]
![image](https://github.com/user-attachments/assets/342f87da-50a6-4ae1-b264-d9e9df1cc148)
not working for examples directory.

**Comment by suryapandian** [comment]
Not having inlay hints is the one reason that is forcing me not to use zed. I have tried toggling `editor::ToggleInlayHints` via command palette adding the setting to settings.json. Nothing works, have also updated zed multiple times. Very frustrating. Also `debug: open language server logs` doesnt show up in my command palette so unfortunately I dont have logs to give more info. 

Anyone out there, any help would be appreciated.

**Reply by biruburu** [reply]
I had the same issue today, and I was able to fix it by explicitly setting `inlay_hints` in my settings.json:

![Screenshot 2025-07-28 at 12 13 58](https://github.com/user-attachments/assets/d33eb4a0-4330-4d43-9092-b544209fc33c)

I did not need to restart, it worked instantly as soon as I added this option. Using zed 0.196.6.

**Reply by joelparkerhenderson** [reply]
Thank you so much! Your solution works for me, and like you say, it works instantly for me too, no need to reload. You've saved my day. <3

**Reply by suryapandian** [reply]
> Not having inlay hints is the one reason that is forcing me not to use zed. I have tried toggling `editor::ToggleInlayHints` via command palette adding the setting to settings.json. Nothing works, have also updated zed multiple times. Very frustrating. Also `debug: open language server logs` doesnt show up in my command palette so unfortunately I dont have logs to give more info.
> 
> Anyone out there, any help would be appreciated.

Apparently I had rust-analyzer 2021 when I bumped it up to 2026 it works 🎉

**Comment by JosephTLyons** [comment]
Thanks for reaching out. We want to better support our community, and the best way for us to do that is through our official Discord. The team is active there and happy to help.

We’re closing the “Help and General Q&A” category on GitHub Discussions and directing all support questions to Discord going forward.

https://zed.dev/community-links
